### PR TITLE
refactor(expr): make evaluation async

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This PR makes all evaluation functions async. It prevents blocking when evaluating UDFs and allows adding blocking pg functions in the future.

In detail, this PR changed the following funcions in trait to be async and refactored all their implementations.

```rust
#[async_trait]
pub trait Expression {
    async fn eval_checked(&self, input: &DataChunk) -> Result<ArrayRef>;
    async fn eval(&self, input: &DataChunk) -> Result<ArrayRef>;
    async fn eval_row(&self, input: &OwnedRow) -> Result<Datum>;
    ...
}

#[async_trait]
pub trait Aggregator {
    async fn update_single(&mut self, input: &DataChunk, row_id: usize) -> Result<()>;
    async fn update_multi(&mut self, input: &DataChunk, start_row_id: usize, end_row_id: usize) -> Result<()>;
    ...
}
```

It's worth notice that introducing async-trait results in an extra memory allocation for each call. It may lead to performance issue when the function is frequently called.

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] ~~I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).~~
- [x] ~~I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)~~
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

I am testing the performance and will show the result later.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

### CODE REVIEW

File: src/expr/src/expr/mod.rs
Comment (jon-chuang) [inline src/expr/src/expr/mod.rs:88]: just curious, is this zero-cost in the case where the function does not contain any actually async blocks that depend on I/O?

Comment (xxchan) [reply src/expr/src/expr/mod.rs:88]: I was also thinking about this problem. Guess it's not 🤪

Comment (xxchan) [reply src/expr/src/expr/mod.rs:88]: btw I found this article. might be interesting read
https://thenewstack.io/using-rustlangs-async-tokio-runtime-for-cpu-bound-tasks/

**Comment by jon-chuang** [comment]
Possibly due to cacheing...? https://www.reddit.com/r/rust/comments/i17arp/rust_proc_macro_attribute_execution_order/

Seems like it should be fine to write a proc_macro within a declarative macro: https://stackoverflow.com/questions/71552272/can-i-execute-a-function-inside-a-declarative-macro

**Comment by jon-chuang** [comment]
Alternate solution if no workaround is found is to implement async_trait desugaring manually in our macros: https://smallcultfollowing.com/babysteps/blog/2023/02/01/async-trait-send-bounds-part-1-intro/

Should not be too hard to achieve.

File: src/expr/src/expr/template.rs
Comment (jon-chuang) [inline src/expr/src/expr/template.rs:31]: concretely, for desugaring async_trait, this function would have to return `Pin<Box<dyn Future + Send>>`

See also: https://blog.rust-lang.org/inside-rust/2022/11/17/async-fn-in-trait-nightly.html

Anw, another option, since we are compiling with nightly anw, is to simply avoid `async_trait` and try `#![feature(async_fn_in_trait)]`? Does it work?

File: src/expr/src/expr/mod.rs
Comment (wangrunji0408) [reply src/expr/src/expr/mod.rs:88]: It's NOT. In fact every async function call in the `async-trait` creates a boxed future, which introduces dynamic memory allocation. It won't be a big problem if it is not called frequently, for example in batch evaluation. But I can't tell how much it will impact on evaluation by row. I'll do a bench for this before merging.

**Comment by wangrunji0408** [comment]
@jon-chuang Thanks for your hint! I'll have a try tomorrow.

File: src/expr/src/expr/template.rs
Comment (wangrunji0408) [reply src/expr/src/expr/template.rs:31]: Manually desugaring seems to be a promising solution. But the async-fn-in-trait feature may not work, because IIRC dynamic dispatch is still in an early stage and not available in nightly.

**Comment by BugenZhao** [comment]
What about keeping the original sync expressions untouched, and introducing a new trait of `AsyncExpression` which is by default implemented for all `Expression`s with a wrapper of `Box::pin`? 🤔

**Comment by jon-chuang** [comment]
>  by default implemented for all Expressions with a wrapper of Box::pin?
I see, so the ones truly requiring async (i.e. just UDF for now) are treated as special cases with manual impl?

Another optimization I thought of is that we can detect if an expression (or subtree) requires async, and build either the sync or async version? Anw, maybe not worth the complexity.

File: src/expr/src/expr/template.rs
Comment (wangrunji0408) [reply src/expr/src/expr/template.rs]: It worked!

**Comment by wangrunji0408** [comment]
> What about keeping the original sync expressions untouched, and introducing a new trait of AsyncExpression which is by default implemented for all Expressions with a wrapper of Box::pin? 🤔

It seems not that easy. As async function is contagious, all `eval` functions will eventually be async as long as they call child functions. Only expressions like InputRef can be kept untouched in this way. 🥹

**Comment by BugenZhao** [comment]
> It seems not that easy. As async function is contagious, all `eval` functions will eventually be async as long as they call child functions. Only expressions like InputRef can be kept untouched in this way. 🥹

You're right. 🥵

File: src/expr/src/expr/expr_some_all.rs
Comment (BugenZhao) [inline src/expr/src/expr/expr_some_all.rs:170]: Not sure whether we should make it concurrent for "left then right"s or "loop"s. But it seems not getting worse considering most of the expressions are CPU bounded.

**Review by BugenZhao — COMMENTED** [review]
Rest LGTM. A great amount of work! 🥵

**Review by TennyZhuang — APPROVED** [review]

**Comment by wangrunji0408** [comment]
Turned it back to draft since there is still a lot of code in the executors to be refactored. 🥵

**Review by xxchan — APPROVED** [review]
😲🥵

**Comment by TennyZhuang** [comment]
Please resolve conflicts

**Comment by wangrunji0408** [comment]
Blocked by simulation test, where the workaround `block_on` for the gap between sync and async code no longer works...

**Comment by BugenZhao** [comment]
> where the workaround `block_on` for the gap between sync and async code no longer works

If we ensure there's no yield point, we can use `now_or_never` as another workaround. 🤔

**Comment by wangrunji0408** [comment]
> > where the workaround `block_on` for the gap between sync and async code no longer works
> 
> If we ensure there's no yield point, we can use `now_or_never` as another workaround. 🤔

Cool idea! thanks!

File: src/frontend/src/expr/mod.rs
Comment (BugenZhao) [inline src/frontend/src/expr/mod.rs:246]: If we need to do this frequently in the future, how about making it an extension to the `Expo` trait?

Comment (wangrunji0408) [reply src/frontend/src/expr/mod.rs:246]: It's okay. But I don't think this should be encouraged. It should only be used in constant evaluation. 🤔

**Comment by wangrunji0408** [comment]
Performance report:

Nexmark throughput
- Q10: -10%
- Q22: -10%
- Q14: +1000% (really weird 🤯)

q10
q14
q22

**Comment by BugenZhao** [comment]
> Q14: +1000% (really weird 🤯)

This should be caused by #8538. 🤣

**Pull request merged**
